pre-conditions.go

```diff
@@ -0,0 +1,89 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage (C) 2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package minio
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+)
+
+// PreConditions - pre conditions implement methods for setting
+// conditions for a GetObject request.
+// http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html
+type PreConditions struct {
+	http.Header
+}
+
+// NewPreConditions - initializes a new pre conditions.
+func NewPreConditions() PreConditions {
+	return PreConditions{
+		Header: make(http.Header),
+	}
+}
+
+// SetMatchETag - set match etag.
+func (c PreConditions) SetMatchETag(etag string) error {
+	if etag == "" {
+		return ErrInvalidArgument("ETag cannot be empty.")
+	}
+	c.Set("If-Match", etag)
+	return nil
+}
+
+// SetMatchETagExcept - set match etag except.
+func (c PreConditions) SetMatchETagExcept(etag string) error {
+	if etag == "" {
+		return ErrInvalidArgument("ETag cannot be empty.")
+	}
+	c.Set("If-None-Match", etag)
+	return nil
+}
+
+// SetUnmodified - set unmodified time since.
+func (c PreConditions) SetUnmodified(modTime time.Time) error {
+	if modTime.IsZero() {
+		return ErrInvalidArgument("Modified since cannot be empty.")
+	}
+	c.Set("If-Unmodified-Since", modTime.Format(http.TimeFormat))
+	return nil
+}
+
+// SetModified - set modified time since.
+func (c PreConditions) SetModified(modTime time.Time) error {
+	if modTime.IsZero() {
+		return ErrInvalidArgument("Modified since cannot be empty.")
+	}
+	c.Set("If-Modified-Since", modTime.Format(http.TimeFormat))
+	return nil
+}
+
+// SetRange - set the start and end offset of the object to be read.
+// See https://tools.ietf.org/html/rfc7233#section-3.1 for reference.
+func (c PreConditions) SetRange(start, end int64) error {
```

Comment:
Yes there are more cases in pure HTTP, But all cases are not supported by AWS S3.